### PR TITLE
fix: show resource kind only when --show-kind is specified

### LINKED ISSUE #669: kubectl parameter `--show-kind` combinded with `-o` sometimes is ignored
`--show-kind=false` does not work when combined with `-o name` 

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get pod
NAME           READY   STATUS    RESTARTS   AGE
chronograf-0   1/1     Running   0          17h
influxdb-0     1/1     Running   0          17h
kapacitor-0    1/1     Running   1          17h

```

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get pod -o wide
NAME           READY   STATUS    RESTARTS   AGE   IP             NODE   NOMINATED NODE   READINESS GATES
chronograf-0   1/1     Running   0          17h   10.244.2.180   zwei   <none>           <none>
influxdb-0     1/1     Running   0          17h   10.244.3.178   drei   <none>           <none>
kapacitor-0    1/1     Running   1          17h   10.244.2.171   zwei   <none>           <none>
```

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get pod -o wide --show-kind=true
NAME               READY   STATUS    RESTARTS   AGE   IP             NODE   NOMINATED NODE   READINESS GATES
pod/chronograf-0   1/1     Running   0          17h   10.244.2.180   zwei   <none>           <none>
pod/influxdb-0     1/1     Running   0          17h   10.244.3.178   drei   <none>           <none>
pod/kapacitor-0    1/1     Running   1          17h   10.244.2.171   zwei   <none>           <none>
```

but

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get pod -o name --show-kind=false
pod/chronograf-0
pod/influxdb-0
pod/kapacitor-0
```

expected:

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get pod -o name --show-kind=false
chronograf-0
influxdb-0
kapacitor-0
```

-------------------

For now I just use formatting with `-o custom-columns=NAME:.metadata.name`, where the kind always is not displayed.

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get po -o custom-columns=NAME:.metadata.name --no-headers=true
chronograf-0
influxdb-0
kapacitor-0
```

and also

```
root@master.k1.cluster.local:~# kubectl -n tick-websites get po -o custom-columns=NAME:.metadata.name --no-headers=true --show-kind=true
chronograf-0
influxdb-0
kapacitor-0
```

### PR DESCRIPTION
**Pull request by hidetatz**
**What type of PR is this?**

/kind bug

**What this PR does / why we need it**:

Currently `kubectl get xx -o name` shows the name with kind. See:

image

However, this is an invalid behavior because `kubectl get` accepts `--show-kind` param which specifies to show kind with the resource name. The expected behavior is:
* when --show-kind is specified, kind is shown
* else, kind is **not** shown

After this PR merged, the problem is fixed:

image

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes/kubectl/issues/669

**Special notes for your reviewer**:

* About changed files

I know there are a lot of changed files which look terrible, but most of them are just tiny change for not making breaking changes.
Basically, there are only 3 parts which actually fixes this bug:

https://github.com/kubernetes/kubernetes/pull/98904/files#diff-646d07042c5cb40b3c11dc4407586c4b3421cb656b7d26c3b88085072adc9af1
https://github.com/kubernetes/kubernetes/pull/98904/files#diff-d6cf14015bc6efb5c381d44f803ba206d37fe0e2910bc6d6bbcb956744442fe3
https://github.com/kubernetes/kubernetes/pull/98904/files#diff-1e4e40e8afb1e2dd3ee9ce305674f709459ae0f6a178ff0a2accaca2d54067ba

We need `WithShowKindOnName` because most kubectl subcommand uses `NamePrintFlags` struct in `PrintFlags`, and they use `NamePrintFlags` to print name. For example, `kubectl edit` command finally shows the message `pod/foo edited` , but the `pod/foo` part is built by `NamePrintFlags`.
Without `WithShowKindOnName` , the message changes to `foo edited`. I personally think it's ok, but I was not very sure if we should make the breaking change, so I chose to explicitly specify to show kind in kubectl subcommands by `WithShowKindOnName`.
If we think the behavior is preferred, then I will remove `WithShowKindOnName` and fix current tests not to show kind.

* unit tests

I ran unit tests on my local machine for `staging/src/k8s.io/cli-runtime` and `staging/src/k8s.io/kubectl` already and made sure they passes.

* Original PR

The original PR was https://github.com/kubernetes/kubernetes/pull/79393. I submitted this PR because the PR seems old , not maintained, and requires some changes.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

### CODE REVIEW

**Comment by BenTheElder** [comment]
> Fixes #669

that seems like the wrong issue.

**Comment by dty1er** [comment]
@BenTheElder 
Ah right, sorry I fixed the PR description.
The correct one is https://github.com/kubernetes/kubectl/issues/669.

**Comment by dty1er** [comment]
Oops, taking a look...

**Comment by dty1er** [comment]
For `pull-kubernetes-node-e2e`, it seems like the flakiness.
e.g. https://github.com/kubernetes/kubernetes/pull/99435#issuecomment-785655965

**Comment by pacoxu** [comment]
> For `pull-kubernetes-node-e2e`, it seems like the flakiness.
> e.g. [#99435 (comment)](https://github.com/kubernetes/kubernetes/pull/99435#issuecomment-785655965)

see #99443

**Comment by dty1er** [comment]
Finally all the tests passed and ready to review

**Review by rikatz — CHANGES_REQUESTED** [review]
Left some small reviews here :)

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags.go [hunk 1]
```diff
@@ -79,5 +83,6 @@ func (f *NamePrintFlags) AddFlags(c *cobra.Command) {}
 func NewNamePrintFlags(operation string) *NamePrintFlags {
 	return &NamePrintFlags{
 		Operation: operation,
+		ShowKind:  false, // by default it doesn't show kind
```
[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags.go:86]: Do you mean we can change this to be true by default? Or we can omit this line?

[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags.go:86]: Ah sorry I got it. I think it's ok to remove it.

File: staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go [hunk 2]
```diff
@@ -232,6 +232,7 @@ func (o *ApplyOptions) Complete(f cmdutil.Factory, cmd *cobra.Command) error {
 	// allow for a success message operation to be specified at print time
 	o.ToPrinter = func(operation string) (printers.ResourcePrinter, error) {
 		o.PrintFlags.NamePrintFlags.Operation = operation
+		o.PrintFlags.NamePrintFlags.ShowKind = true
```
[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go:235]: Ah right. will remove

[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/kubectl/pkg/cmd/apply/apply.go]: fixed: https://github.com/kubernetes/kubernetes/pull/98904/commits/7597df4966a6b479de10d242f5d11af8e8c844ec

File: staging/src/k8s.io/cli-runtime/pkg/printers/name.go [hunk 3]
```diff
@@ -121,10 +124,18 @@ func printObj(w io.Writer, name string, operation string, shortOutput bool, grou
...
+		}
 		return nil
 	}
 
-	fmt.Fprintf(w, "%s.%s/%s%s\n", strings.ToLower(groupKind.Kind), groupKind.Group, name, operation)
+	if showKind {
+		fmt.Fprintf(w, "%s.%s/%s%s\n", strings.ToLower(groupKind.Kind), groupKind.Group, name, operation)
+	} else {
+		fmt.Fprintf(w, "%s%s\n", name, operation)
+	}
```
[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/printers/name.go]: Thank you for the suggestion!
I fixed it based on your suggestion; I changed it to use 2 vars for more clarity
https://github.com/kubernetes/kubernetes/pull/98904/commits/28763c4eb22be71251f25578e69f793dc9e83406

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/print_flags_test.go [hunk 4]
```diff
@@ -20,7 +20,7 @@ import (
 	"bytes"
 	"testing"
 
-	"k8s.io/api/core/v1"
+	v1 "k8s.io/api/core/v1"
```
[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/print_flags_test.go]: Thank you, fixed: https://github.com/kubernetes/kubernetes/pull/98904/commits/5031dc2c195791505e44f34e1589ebee5c2eacc5

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags.go]: fixed: https://github.com/kubernetes/kubernetes/pull/98904/commits/d965fcbe2b15b2ad02f83ccb30f908e16d65b7c9

File: staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags_test.go [hunk 5]
```diff
@@ -43,33 +44,45 @@ func TestNamePrinterSupportsExpectedFormats(t *testing.T) {
 		{
 			name:           "valid \"name\" output format with no operation prints resource name",
 			outputFormat:   "name",
+			showKind:       false,
```
[the remark this answers is not in the document]
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/genericclioptions/name_flags_test.go]: fixed: https://github.com/kubernetes/kubernetes/pull/98904/commits/d965fcbe2b15b2ad02f83ccb30f908e16d65b7c9

**Comment by dty1er** [comment]
@rikatz 
Thank you for your review. I applied some fixes based on your comment. Can you take a look again?

File: staging/src/k8s.io/cli-runtime/pkg/printers/name_test.go [hunk 6]
```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2019 The Kubernetes Authors.
```
Comment (rikatz) [inline staging/src/k8s.io/cli-runtime/pkg/printers/name_test.go:2]: Nit here: As you're creating a new file, maybe change the year to 2021 :)

**Comment by rikatz** [comment]
@dty1er thank you!

Left a small nit (sorry, didn't saw on my first pass) but otherwise this looks good to me.

Will wait someone else to take a final look :)

File: staging/src/k8s.io/cli-runtime/pkg/printers/name_test.go (hunk 6 above)
Comment (dty1er) [reply staging/src/k8s.io/cli-runtime/pkg/printers/name_test.go]: fixed: https://github.com/kubernetes/kubernetes/pull/98904/commits/8c42efdd47040c7855f3807d22b82696e4094e30

**Comment by liggitt** [comment]
/hold

isn't changing the output of `-o name` like this a breaking change? the output can currently be piped to `xargs kubectl get` and is well-defined

cc @deads2k

File: test/cmd/crd.sh [hunk 7]
```diff
@@ -184,15 +184,15 @@ run_non_native_resource_tests() {
   # Test that we can create a new resource of type Kind
   kubectl "${kube_flags[@]}" create -f hack/testdata/CRD/resource.yaml "${kube_flags[@]}"
 
-  # Test that -o name returns kind.group/resourcename
+  # Test that -o name returns only resourcename
   output_message=$(kubectl "${kube_flags[@]}" get resource/myobj -o name)
-  kube::test::if_has_string "${output_message}" 'kind.mygroup.example.com/myobj'
+  kube::test::if_has_string "${output_message}" 'myobj'
```
Comment (liggitt) [inline test/cmd/crd.sh:189]: needing to change tests like this is an excellent sign that this is breaking compatibility

**Pull request closed without merging**